### PR TITLE
Find the stemcell with specifying filter

### PR DESCRIPTION
**Pull request by mattcui**
Softlayer provides object filter feature to help precisely identify objects, with this improvement, softlayer could return only one stemcell back, so that we don't need the code to find the tareget stemcell by trying to match id or uuid in a loop. There would be hundreds or thousands stemcells under an account. So there probably returns "timed-out" error sometimes if we retrieve all stemcells from SL. It's a lesson learned from legacy Ruby CPI.

Reference:
http://sldn.softlayer.com/article/Object-Filters

### CODE REVIEW

**Comment by maximilien** [comment]
You need to resolve conflicts first. Likely you need to rebase. 

Please update and when you do I'll review. Thanks.

**Comment by mattcui** [comment]
@maximilien I fetched and merged the latest code from master, please help review and merge. Thanks.

**Comment by maximilien** [comment]
@mattcui, you still have conflicts. See attached pic of what I see on github

find_the_stemcell_with_specifying_filter_by_mattcui_ _pull_request__51_ _maximilien_bosh-softlayer-cpi

File: softlayer/stemcell/softlayer_finder_test.go [hunk 1]
```diff
@@ -49,4 +49,26 @@ var _ = Describe("SoftLayerFinder", func() {
...
+				Expect(err).ToNot(HaveOccurred())
+				Expect(found).To(BeTrue())
+
+				expectedStemcell := NewSoftLayerStemcell(4868344, "8c7a8358-d9a9-4e4d-9345-6f637e10ccb7", VirtualDiskImageKind, softLayerClient, logger)
+				Expect(stemcell).To(Equal(expectedStemcell))
+			})
+		})
+
+		Context("valid stemcell ID pointing to a SL virtual disk image", func() {
+			It("returns found as false if stemcell directory does not exist", func() {
```
Comment (maximilien) [inline softlayer/stemcell/softlayer_finder_test.go]: can we fix the `It` string to be better. Seems that "stemcell directory does not exist" is not correct description...

File: softlayer/stemcell/softlayer_finder_test.go [hunk 2]
```diff
@@ -49,4 +49,26 @@ var _ = Describe("SoftLayerFinder", func() {
...
 	})
+
+	Describe("FindById", func() {
+		Context("valid stemcell ID pointing to a SL virtual disk image", func() {
+			It("returns stemcell and found as true if stemcell ", func() {
+				stemcell, found, err := finder.FindById(4868344)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(found).To(BeTrue())
+
+				expectedStemcell := NewSoftLayerStemcell(4868344, "8c7a8358-d9a9-4e4d-9345-6f637e10ccb7", VirtualDiskImageKind, softLayerClient, logger)
```
Comment (maximilien) [inline softlayer/stemcell/softlayer_finder_test.go]: Usually we move expected objects in `BeforeEach` statements or at the very least at the start of the `It` statement. This is just a matter of style (for consistency).

Also, I'd check and compare the `Id` and Uuid` of the two objects?

File: softlayer/stemcell/softlayer_finder_test.go [hunk 3]
```diff
@@ -40,7 +40,7 @@ var _ = Describe("SoftLayerFinder", func() {
 			})
 		})
 
-		Context("valid stemcell ID pointing to a SL virtual disk image", func() {
+		Context("valid stemcell UUID pointing to a SL virtual disk image", func() {
 			It("returns found as false if stemcell directory does not exist", func() {
```
Comment (maximilien) [inline softlayer/stemcell/softlayer_finder_test.go]: Same here. Let's make the description better, use something like:

`"returns found as false if stemcell directory does not exist"`

Thanks

**Comment by maximilien** [comment]
BTW, this is looking good. Once you rebase and fix the minor updates I suggested, then I will merge.

Thanks.

File: softlayer/stemcell/softlayer_finder_test.go (hunk 3 above)
Comment (mattcui) [inline softlayer/stemcell/softlayer_finder_test.go]: Copied from the original Find test function, updated both functions.

File: softlayer/stemcell/softlayer_finder_test.go (hunk 1 above)
Comment (mattcui) [inline softlayer/stemcell/softlayer_finder_test.go]: Copied from the original Find test function, updated both functions.

File: softlayer/stemcell/softlayer_finder_test.go (hunk 2 above)
Comment (mattcui) [inline softlayer/stemcell/softlayer_finder_test.go]: Moved `expectedStemcell` in `BeforeEach` function.

What do you mean by "I'd check and compare the Id and Uuid of the two objects?" Thanks.

**Comment by mattcui** [comment]
@maximilien Updated according to your comments, and fetched/merged again. Please help review, thanks.

File: softlayer/stemcell/softlayer_finder_test.go [hunk 4]
```diff
@@ -26,23 +27,43 @@ var _ = Describe("SoftLayerFinder", func() {
...
+				Expect(found).To(BeTrue())
 				Expect(stemcell).To(Equal(expectedStemcell))
 			})
 		})
 
 		Context("valid stemcell ID pointing to a SL virtual disk image", func() {
-			It("returns found as false if stemcell directory does not exist", func() {
-				stemcell, found, err := finder.Find("fake-stemcell-id")
+			It("returns found as false if stemcell does not exist", func() {
+				stemcell, found, err := finder.FindById(1111111)
```
Comment (maximilien) [inline softlayer/stemcell/softlayer_finder_test.go]: No new line needed between `Expect` calls that are testing the same objects. Typically we just group related assertions.

**Comment by maximilien** [comment]
Nice. Minor comment but merging anyhow, feel free to address comment in future commits.

Cheers :beers:

**Pull request merged**